### PR TITLE
[Admin] Adding tax rate feature

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        And the store has a tax category "Food and Beverage" with a code "food_and_beverage"
+        And the store has a default calculator defined
+        And I am logged in as an administrator
+
+    @todo
+    Scenario: Adding a new tax rate
+        Given I want to create a new tax rate
+        When I specify its code as "EUROPEAN_UNION_SALES_TAX"
+        And I name it "European Union Sales Tax"
+        And It is defined for EU zone
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: It -> it

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        And It defines tax rates for "Food and Beverage" tax category
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: It -> it

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        And I its amount is "20%"
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: `its amount is "20%"`?

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        And I choose default calculator
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: Possibly **the** default calculator? /cc @TheMadeleine

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        And I am logged in as an administrator
+
+    @todo
+    Scenario: Adding a new tax rate
+        Given I want to create a new tax rate
+        When I specify its code as "EUROPEAN_UNION_SALES_TAX"
+        And I name it "European Union Sales Tax"
+        And it is defined for EU zone
+        And it defines tax rates for "Food and Beverage" tax category
+        And its amount is "20%"
```
Comment (tuka217) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: `And its`

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        And the store has a tax category "Food and Beverage" with a code "food_and_beverage"
+        And the store has a default calculator defined
+        And I am logged in as an administrator
+
+    @todo
+    Scenario: Adding a new tax rate
+        Given I want to create a new tax rate
+        When I specify its code as "EUROPEAN_UNION_SALES_TAX"
+        And I name it "European Union Sales Tax"
+        And it is defined for a EU zone
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: Anyway, we use "I specify its code", "I name it" but the "it is defined", "it defines", looks inconsistent.

Shouldn't it be "I define it for the EU zone", "I make it applicable for "Food and Beverage" tax category", etc.?

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 7]
```diff
@@ -0,0 +1,24 @@
+@managing_tax_rates
+Feature: Adding a new tax rate
+    In order to apply different taxes rates to various tax categories
+    As an Administrator
+    I want to add a new tax rate to the registry
+
+    Background:
+        Given there is "EU" zone containing all members of European Union
+        And the store has a tax category "Food and Beverage" with a code "food_and_beverage"
+        And the store has a default calculator defined
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: default **tax** calculator?

File: features/taxation/managing_tax_rates/adding_tax_rate.feature (hunk 6 above)
Comment (lchrusciel) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: Thats was what I needed

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 8]
```diff
@@ -0,0 +1,24 @@
+@managing_tax_rates
+Feature: Adding a new tax rate
+    In order to apply different taxes rates to various tax categories
+    As an Administrator
+    I want to add a new tax rate to the registry
+
+    Background:
+        Given there is "EU" zone containing all members of European Union
+        And the store has a tax category "Food and Beverage" with a code "food_and_beverage"
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: Can we get rid of the tax category code? It does not have any meaning in this feature.

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 9]
```diff
@@ -0,0 +1,24 @@
+@managing_tax_rates
+Feature: Adding a new tax rate
+    In order to apply different taxes rates to various tax categories
+    As an Administrator
+    I want to add a new tax rate to the registry
+
+    Background:
+        Given there is "EU" zone containing all members of European Union
```
Comment (pamil) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: > there is "EU" zone
> it is defined for a EU zone

Let's either use the quotation marks or not :)

**Comment by pamil** [comment]
Looks good! :+1:

File: features/taxation/managing_tax_rates/adding_tax_rate.feature (hunk 9 above)
Comment (lchrusciel) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: :+1: Applicated

Comment (TheMadeleine) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: `Given there is a zone "EU" containing all members of the European Union`

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        And the store has a tax category "Food and Beverage"
+        And the store has a default tax calculator defined
+        And I am logged in as an administrator
+
+    @todo
+    Scenario: Adding a new tax rate
+        Given I want to create a new tax rate
+        When I specify its code as "EUROPEAN_UNION_SALES_TAX"
+        And I name it "European Union Sales Tax"
+        And I define it for a "EU" zone
```
Comment (TheMadeleine) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: for the "EU" zone.
 Cause from now on it is already defined in the system and we know which one it is. :D

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        And I make it applicable for "Food and Beverage" tax category
```
Comment (TheMadeleine) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: `the "Food and Beverage" tax category`

File: src/Sylius/Behat/Context/Transform/TaxRateContext.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * @Transform :taxRate
+     */
+    public function getTaxRateByName($taxRateName)
+    {
+        $taxRate = $this->taxRateRepository->findOneBy(['name' => $taxRateName]);
+        if (null === $taxRate) {
+            throw new \InvalidArgumentException(sprintf('Tax category with name "%s" does not exist', $taxRateName));
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Transform/TaxRateContext.php]: Missing dot at the end of exception message.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+     */
+    private $notificationAccessor;
+
+    /**
+     * @param IndexPageInterface $indexPage
+     * @param CreatePageInterface $createPage
+     * @param UpdatePageInterface $updatePage
+     * @param NotificationAccessorInterface $notificationAccessor
+     */
+    public function __construct(IndexPageInterface $indexPage, CreatePageInterface $createPage, UpdatePageInterface $updatePage, NotificationAccessorInterface $notificationAccessor)
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: This page is quite long, we should split it.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+        $this->createPage->create();
+    }
+
+    /**
+     * @Then the tax rate :taxRateName should appear in the registry
+     */
+    public function theTaxRateShouldAppearInTheRegistry($taxRateName)
+    {
+        $this->indexPage->open();
+        Assert::true($this->indexPage->isResourceOnPage(['name' => $taxRateName]), sprintf('Tax rate with name %s has not been found.', $taxRateName));
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: 1. Too long line.
2. I would add blank line before that.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+     * @param string $taxRateElement
+     */
+    private function assertFieldValue(TaxRateInterface $taxRate, $element, $taxRateElement)
+    {
+        $this->indexPage->open();
+        Assert::true(
+            $this->indexPage->isResourceOnPage(
+                [
+                    'code' => $taxRate->getCode(),
+                    $element => $taxRateElement
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: Missing comma at the end.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+                ]
+            ),
+            sprintf('Tax rate %s %s was not assigned properly.', $element, $taxRateElement)
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: Wdyt about ``'Tax rate %s %s has not been assigned properly.'``?

File: src/Sylius/Behat/Behaviour/ChoosesCalculator.php [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Behaviour;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+trait ChoosesCalculator
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Behaviour/ChoosesCalculator.php]: We do not have many other calculators, I think in this case trait is an overkill.

File: src/Sylius/Behat/Behaviour/ChoosesCategory.php [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Behaviour;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+trait ChoosesCategory
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Behaviour/ChoosesCategory.php]: Same for this trait.

File: src/Sylius/Behat/Behaviour/ChoosesZone.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Behaviour;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+trait ChoosesZone
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Behaviour/ChoosesZone.php]: I'd skip this one as well.

File: src/Sylius/Behat/Behaviour/SpecifiesItsAmount.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Behaviour;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+trait SpecifiesItsAmount
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Behaviour/SpecifiesItsAmount.php]: Same.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+    {
+        $this->updatePage->saveChanges();
+    }
+
+    /**
+     * @Then I should be notified about successful edition
+     */
+    public function iShouldBeNotifiedAboutSuccessfulEdition()
+    {
+        $this->updatePage->isCodeDisabled();
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: Redundant

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+
+    /**
+     * @param TaxRateInterface $taxRate
+     * @param string $element
+     * @param string $taxRateElement
+     */
+    private function assertFieldValue(TaxRateInterface $taxRate, $element, $taxRateElement)
+    {
+        $this->indexPage->open();
+        Assert::true(
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: Missing blank line

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing/tax_rate.yml [hunk 23]
```diff
@@ -0,0 +1,16 @@
+sylius_admin_tax_category:
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing/tax_rate.yml]: `sylius_admin_tax_rate`

File: src/Sylius/Behat/Context/Transform/TaxRateContext.php [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * @Transform :taxRate
+     */
+    public function getTaxRateByName($taxRateName)
+    {
+        $taxRate = $this->taxRateRepository->findOneBy(['name' => $taxRateName]);
+        if (null === $taxRate) {
+            throw new \InvalidArgumentException(sprintf('Tax category with name "%s" does not exist.', $taxRateName));
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Transform/TaxRateContext.php]: rate

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+    /**
+     * @Then there should still be only one tax rate with :element :code
+     */
+    public function thereShouldStillBeOnlyOneTaxRateWith($element, $code)
+    {
+        $this->indexPage->open();
+
+        Assert::true(
+            $this->indexPage->isResourceOnPage([$element => $code]),
+            sprintf('Tax category with %s %s cannot be founded.', $element, $code)
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: rate

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+
+    @ui
+    Scenario: Adding a new tax rate
+        Given I want to create a new tax rate
+        When I specify its code as "EUROPEAN_UNION_SALES_TAX"
+        And I name it "European Union Sales Tax"
+        And I define it for the "European Union" zone
+        And I make it applicable for the "Food and Beverage" tax category
+        And I specify its amount as 20%
+        And I want to use the default tax calculator
```
Comment (michalmarcinkowski) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature]: And I ~~want to use~~ **choose**?

File: src/Sylius/Bundle/AdminBundle/Resources/views/TaxRate/_form.html.twig [hunk 27]
```diff
@@ -0,0 +1,9 @@
+{{ form_errors(form) }}
+<div class="three fields">
+    {{ form_row(form.code) }}
+    {{ form_row(form.name) }}
+    {{ form_row(form.amount) }}
+</div>
+{{ form_row(form.category) }}
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/views/TaxRate/_form.html.twig]: Can be placed in 3 rows as well

**Comment by michalmarcinkowski** [comment]
Nice work! :+1:

File: features/taxation/managing_tax_rates/editing_tax_rate.feature [hunk 28]
```diff
@@ -0,0 +1,58 @@
+@managing_tax_rates
+Feature: Editing tax rate
+    In order to change tax classification of my products
```
Comment (pjedrzejewski) [inline features/taxation/managing_tax_rates/editing_tax_rate.feature]: In order to change taxes applied to orders

Comment (michalmarcinkowski) [inline features/taxation/managing_tax_rates/editing_tax_rate.feature]: "taxes applied to **products**" will be more appropriate for me :)

**Pull request merged**